crates/bevy_reflect/src/func/mod.rs

```diff
@@ -34,164 +33,107 @@
 //! assert_eq!(value.unwrap_owned().downcast_ref::<i32>(), Some(&100));
 //! ```
 //!
+//! # Functions vs Closures
+//!
+//! In Rust, a "function" is any callable that does not capture its environment.
+//! These are typically defined with the `fn` keyword, but may also use anonymous function syntax.
+//!
+//! ```rust
+//! // This is a standard Rust function:
+//! fn add(a: i32, b: i32) -> i32 {
+//!   a + b
+//! }
+//!
+//! // This is an anonymous Rust function:
+//! let add = |a: i32, b: i32| a + b;
+//! ```
+//!
+//! Rust also has the concept of "closures", which are special functions that capture their environment.
+//! These are always defined with anonymous function syntax.
+//!
+//! ```rust
+//! // A closure that captures an immutable reference to a variable
+//! let c = 123;
+//! let add = |a: i32, b: i32| a + b + c;
+//!
+//! // A closure that captures a mutable reference to a variable
+//! let mut total = 0;
+//! let add = |a: i32, b: i32| total += a + b;
+//!
+//! // A closure that takes ownership of its captured variables by moving them
+//! let c = 123;
+//! let add = move |a: i32, b: i32| a + b + c;
+//! ```
+//!
+//! Each callable may be considered a subset of the other:
+//! functions are a subset of immutable closures which are a subset of mutable closures.
+//!
+//! This means that, in terms of traits, you could imagine that any type that implements
+//! [`IntoFunction`], also implements [`IntoClosure`] and [`IntoClosureMut`].
+//! And every type that implements [`IntoClosure`] also implements [`IntoClosureMut`].
+//!
+//! # Valid Signatures
+//!
+//! Many of the traits in this module have default blanket implementations over a specific set of function signatures.
+//!
+//! These signatures are:
+//! - `fn(...) -> R`
```

Comment:
As I commented elsewhere, I would avoid naming the functions `fn` unless these are supposed to be the types of function pointers